includes/Classifai/Providers/Azure/Personalizer.php

```diff
@@ -0,0 +1,604 @@
+			$class .= ' has-author';
+		}
+		return sprintf(
+			'<ul class="%1$s">%2$s</ul>',
+			esc_attr( $class ),
+			$markup
+		);
+	}
+
+	/**
+	 * Get Features for specific post.
+	 *
+	 * @param int $post_id Post Object.
+	 * @return array
+	 */
+	protected function get_post_features( $post_id ) {
+		$features = array(
+			'title'   => $this->get_string_words( get_the_title( $post_id ) ),
+			'excerpt' => $this->get_string_words( get_the_excerpt( $post_id ) ),
+			'_URL'    => get_the_permalink( $post_id ),
+		);
+
+		$post_type = get_post_type( $post_id );
+
+		// Get post type taxonomies.
+		$taxonomies = get_object_taxonomies( $post_type, 'objects' );
+		foreach ( $taxonomies as $taxonomy_slug => $taxonomy ) {
+			// Get the terms related to post.
+			$terms = get_the_terms( $post_id, $taxonomy_slug );
+
+			if ( ! empty( $terms ) ) {
+				foreach ( $terms as $term ) {
+					$features[ $taxonomy->label ][ $term->name ] = 1;
+				}
+			}
+		}
+
+		return $features;
+	}
+
+	/**
+	 * Get user agent for personilizer contextFeatures.
+	 */
+	protected function get_user_agent_features() {
+		// phpcs:ignore WordPressVIPMinimum.Variables.RestrictedVariables.cache_constraints___SERVER__HTTP_USER_AGENT__
+		$user_agent = isset( $_SERVER['HTTP_USER_AGENT'] ) ? sanitize_text_field( wp_unslash( $_SERVER['HTTP_USER_AGENT'] ) ) : '';
+
+		// User Agent Parsing
+		$parser    = Parser::create();
+		$ua_result = $parser->parse( $user_agent );
+
+		return array(
+			'_ua'           => $ua_result->originalUserAgent, // phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+			'_DeviceFamily' => $ua_result->device->family,
+			'_OSFamily'     => $ua_result->os->family,
+			'DeviceBrand'   => $ua_result->device->brand,
+			'DeviceModel'   => $ua_result->device->model,
+		);
+	}
+
+	/**
+	 * Get array of words from string. Words as key of array.
+	 *
+	 * @param string $string String to get words from.
+	 * @return array
+	 */
+	protected function get_string_words( $string ) {
+		$str_array = preg_split( '/\s+/', $string );
+		$words     = array();
+		foreach ( $str_array as $str ) {
+			$words[ $str ] = 1;
+		}
+
+		return $words;
+	}
+
+	/**
+	 * Get Ranked action by sending request to Azure personalizer.
+	 *
+	 * @param array $rank_request Prepared Request data.
+	 * @return object|string
+	 */
+	protected function personalizer_get_ranked_action( $rank_request ) {
+		$settings = $this->get_settings();
+		$result   = wp_remote_post(
+			trailingslashit( $settings['url'] ) . $this->rank_endpoint,
+			[
+				'headers' => [
+					'Ocp-Apim-Subscription-Key' => $settings['api_key'],
+					'Content-Type'              => 'application/json',
+				],
+				'body'    => wp_json_encode( $rank_request ),
+			]
+		);
+
+		if ( ! is_wp_error( $result ) ) {
+			$response = json_decode( wp_remote_retrieve_body( $result ) );
+			set_transient( 'classifai_azure_personalizer_rank_response', $response, DAY_IN_SECONDS * 30 );
```

Comment:
So we cache the response here but doesn't appear we ever use this?

Reply:
Yes, nice catch. This is leftover code, it was initially added to show it in site health info. but later I displayed service status info.